specs/error-handling-guideline.md

```diff
@@ -0,0 +1,68 @@
+
+Last, error logs can also be useful for post-mortem debugging and can also be written to a file, truncate them occasionally and make sure they don't contain ansi color codes.
+
+## Error output recommendation
+
+### Dos
+
+- Provide full description if the user input does not match what Notation CLI expected. A full description should include the actual input received from the user and expected input.
+- Use the capital letter ahead of each line of any error message.
+- Print human readable error message. If the error message is mainly from the server and varies by different servers, tell users that the error response is from server. This implies that users may need to contact server side for troubleshooting.
+- Provide specific and actionable prompt message with argument suggestion or show the example usage for reference. (e.g, Instead of showing flag or argument options is missing, please provide available argument options and guide users to `--help` to view more examples).
+- If the actionable prompt message is too long to show in the CLI output, consider guide users to Notation user manual or troubleshooting guide with the versioned permanent link.
+- If the error message is not enough for troubleshooting, guide users to use `--verbose` to print much more detailed logs.
+- If server returns an error without any [message or detail](https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc.3/spec.md#error-codes), consider providing customized error logs to make it clearer. The original server logs can be displayed in debug mode.
+- As a security tool, `notation` SHOULD prompt users to stop upon verification errors. 
+
+### Don'Ts
+
+- Do not use a formula-like or a programming expression in the error message. (e.g, `json: cannot unmarshal string into Go value of type map[string]map[string]string.`, or `Parameter 'xyz' must conform to the following pattern: '^[-\\w\\._\\(\\)]+$'`).
+- Do not use ambiguous expressions which mean nothing to users. (e.g, `Something unexpected happens`, or `Error: accepts 2 arg(s), received 0`).
+- Do not print irrelevant error message to make the output noisy. The more irrelevant output you produce, the longer it's going to take the user to figure out what they did wrong.
+- As a security tool, `notation` MUST NOT prompt users to update the trust policy to bypass the verification error when verification fails.
+- Key and certificate information used in the verification are considered sensitive so it's not recommended to print them out. 
+
+## How to write friendly error message
+
+### Recommended error message structure
+
+Here is a sample structure of an error message:
+
+```text
+{Error|Error response from a remote registry or a server}: {Error description (HTTP status code can be printed out if any)}
```

Comment:
The error message from a remote server usually includes an HTTP status code. If the error message contains the HTTP status code, do we still need to mention that it is an `Error response from a remote registry or server`? Is it redundant?

Reply:
I find that not all error messages contain HTTP status code. For example:

```
notation login localhost:5000/
Username: test
Password: 
Error: failed to get registry client: invalid reference: invalid registry
```

Reply:
In this example, the error was returned from CLI directly without requesting server.

Reply:
I think there is no conflict with printing HTTP status code and "Error response from a remote registry or a server" because not all users are able to understand the meaning of HTTP status code. The descriptive error message prompts users that they can consult the server side for further help or investigation.